QualifiedTeachersApi/src/QualifiedTeachersApi/Jobs/BatchSendQtsAwardedEmailsJob.cs

```diff
@@ -0,0 +1,81 @@
+﻿using Microsoft.EntityFrameworkCore;
+using Microsoft.Extensions.Options;
+using QualifiedTeachersApi.DataStore.Crm;
+using QualifiedTeachersApi.DataStore.Sql;
+using QualifiedTeachersApi.DataStore.Sql.Models;
+using QualifiedTeachersApi.Jobs.Scheduling;
+
+namespace QualifiedTeachersApi.Jobs;
+
+public class BatchSendQtsAwardedEmailsJob
+{
+    private readonly BatchSendQtsAwardedEmailsJobOptions _batchSendQtsAwardedEmailsJobOptions;
+    private readonly DqtContext _dbContext;
+    private readonly IDataverseAdapter _dataverseAdapter;
+    private readonly IBackgroundJobScheduler _backgroundJobScheduler;
+    private readonly IClock _clock;
+
+    public BatchSendQtsAwardedEmailsJob(
+        IOptions<BatchSendQtsAwardedEmailsJobOptions> batchSendQtsAwardedEmailsJobOptions,
+        DqtContext dbContext,
+        IDataverseAdapter dataverseAdapter,
+        IBackgroundJobScheduler backgroundJobScheduler,
+        IClock clock)
+    {
+        _batchSendQtsAwardedEmailsJobOptions = batchSendQtsAwardedEmailsJobOptions.Value;
+        _dbContext = dbContext;
+        _dataverseAdapter = dataverseAdapter;
+        _backgroundJobScheduler = backgroundJobScheduler;
+        _clock = clock;
+    }
+
+    public async Task Execute(CancellationToken cancellationToken)
+    {
+        var lastAwardedToUtc = _batchSendQtsAwardedEmailsJobOptions.InitialLastAwardedToUtc;
+        var lastExecutedJob = await _dbContext.QtsAwardedEmailsJobs.OrderBy(j => j.ExecutedUtc).LastOrDefaultAsync();
+        if (lastExecutedJob != null)
+        {
+            lastAwardedToUtc = lastExecutedJob.AwardedToUtc;
+        }
+
+        // Look for new QTS awards up to the end of the day the configurable amount of days ago to provide a delay between award being given and email being sent.
+        var awardedToUtc = (_clock.Today.ToDateTime() - TimeSpan.FromDays(_batchSendQtsAwardedEmailsJobOptions.EmailDelayDays)).AddDays(1) - TimeSpan.FromSeconds(1);
+
+        var executed = _clock.UtcNow;
+        var startDate = lastAwardedToUtc.AddSeconds(1);
+        var endDate = awardedToUtc;
+        var qtsAwardedEmailsJobId = Guid.NewGuid();
+        var job = new QtsAwardedEmailsJob
+        {
+            QtsAwardedEmailsJobId = qtsAwardedEmailsJobId,
+            AwardedToUtc = awardedToUtc,
+            ExecutedUtc = executed
+        };
+        await _dbContext.QtsAwardedEmailsJobs.AddAsync(job, cancellationToken);
+        await _dbContext.SaveChangesAsync(cancellationToken);
```

Comment:
It would be good to put this whole block inside a `TransactionScope` (including the `_backgroundJobScheduler.Enqueue`)

Reply:
Not sure what this gives us i.e. which bits are atomic?
The job which processes the emails get the job item details from the database so if that is in another transaction it won't be in the database as far as that job goes

Reply:
and passing a transaction around across background jobs doesn't feel quite right

Reply:
Sorted this now